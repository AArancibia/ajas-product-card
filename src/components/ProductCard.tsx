import React from "react";
import { createContext, CSSProperties } from "react";
import styles from '../styles/styles.module.css';

import { useProduct } from "../hooks/useProduct";
import {ProductContextProps,Product, onChangeArgs, InitialValues, ProductCardHandlers} from '../interfaces/interfaces';


export const ProductContext = createContext({} as ProductContextProps);
const {Provider} = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({children, product, value, className, style, onChange, initialValues}: Props) => {
    
    const {counter, increaseBy, maxCount, isMaxCountReached, reset} = useProduct({onChange, product, value, initialValues});

    return (
      <Provider value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}>
        <div 
          className={`${styles.productCard} ${className}`}
          style={style}
        >
          {
            children({
              count: counter,
              isMaxCountReached,
              maxCount: initialValues?.count,
              increaseBy,
              reset,
              product,
            })
          }
        </div>
      </Provider>
    );
}


