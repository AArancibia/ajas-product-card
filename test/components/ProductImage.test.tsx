import React from "react";
import { ProductImage, ProductCard } from "../../src/components";
import renderer from "react-test-renderer";
import { product2 } from "../data/products";

describe('ProductImage', () => {

    test('should show the component with a correct title', () => {
        const wrapper = renderer.create(
            <ProductImage img="http://alexis.com"></ProductImage>
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
      });

      
    
    test('show show the right image for product', () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                () => (<ProductImage></ProductImage>)
                }
            </ProductCard>
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    })
})