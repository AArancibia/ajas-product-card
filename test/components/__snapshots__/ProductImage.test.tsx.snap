// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show the component with a correct title 1`] = `
<img
  alt="Product image"
  className="productImg undefined"
  src="http://alexis.com"
/>
`;

exports[`ProductImage show show the right image for product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product image"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
